src/hal/src/pso/descriptor.rs

```diff
@@ -67,6 +67,91 @@ pub enum DescriptorType {
     InputAttachment = 10,
 }
 
+impl std::convert::From<DescriptorType> for RawDescriptorType {
+    fn from(ty: DescriptorType) -> Self {
+        match ty {
+            DescriptorType::Sampler => RawDescriptorType::Sampler,
+            DescriptorType::CombinedImageSampler => RawDescriptorType::CombinedImageSampler,
+            DescriptorType::Image { ty } => match ty {
+                ImageDescriptorType::Storage => RawDescriptorType::StorageImage,
+                ImageDescriptorType::Sampled => RawDescriptorType::SampledImage,
+            }
+            DescriptorType::Buffer { access, format } => match access {
+                BufferDescriptorAccess::Storage => match format {
+                    BufferDescriptorFormat::Dynamic => RawDescriptorType::StorageBufferDynamic,
+                    BufferDescriptorFormat::Structured => RawDescriptorType::StorageBuffer,
+                    BufferDescriptorFormat::Texel => RawDescriptorType::StorageTexelBuffer,
+                },
+                BufferDescriptorAccess::Uniform => match format {
+                    BufferDescriptorFormat::Dynamic => RawDescriptorType::UniformBufferDynamic,
+                    BufferDescriptorFormat::Structured => RawDescriptorType::UniformBuffer,
+                    BufferDescriptorFormat::Texel => RawDescriptorType::UniformTexelBuffer,
+                },
+            }
+            DescriptorType::InputAttachment => RawDescriptorType::InputAttachment,
+        }
+    }
+}
+
+/// Access type of a buffer.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub enum BufferDescriptorAccess {
+    /// Storage buffers allow load, store, and atomic operations.
+    Storage,
+    /// Uniform buffers allow only load operations.
+    Uniform,
+}
+
+/// Format of a buffer.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub enum BufferDescriptorFormat {
+    // TODO: seems to have no effect in hal, ask about removing or documenting that it's a no-op
+    /// TODO
+    Dynamic,
+    /// The buffer is interpreted as a structure defined in a shader.
+    Structured,
```

Comment:
instead of having a separate `Dynamic` variant here, let's have this as a bool in
```rs
Structured { dynamic_offset: bool },
```